### PR TITLE
Si 6.39 (Беклог №8) Написать тесты на игровой движок + любой UI

### PR DESCRIPTION
**Pull request by IrinaSha**
### Какую задачу решаем
Покрытие тестами движка игры
 src/game/engine   
  Assets.ts              
  Game.ts  
  Input.ts 
  Level.ts              
  types.ts  
 src/game/entities                
  Bullet.ts  
  Explosion.ts                          
  Fleet.ts                                 
  Player.ts        
  Shield.ts 
        
 src/pages/profile                          
  GeoSection.tsx               
  PasswordSection.tsx  
  ProfilePage.tsx
  ProfileSection.tsx               
  ProfileService.ts

### TBD (если есть)

### CODE REVIEW

File: packages/client/src/game/engine/Game.test.ts [hunk 1]
```diff
@@ -258,11 +352,11 @@ describe('Game', () => {
   it('handleUiClick ignores null button and misses', () => {
     const { game } = mkGame()
     game.uiButton = null
-    game.handleUiClick(400, 400)
+    game.handleUiClick(UI_CLICK_HIT_X, UI_CLICK_HIT_Y)
```
Comment (Antonov2504) [inline packages/client/src/game/engine/Game.test.ts]: Думаю, что тесты на клик по канвасу можно убрать, т.к. в своем PR удалил оверлей канваса и связал работу движка игры с модальным окном реакта

File: packages/client/src/pages/profile/consts.ts [hunk 2]
```diff
@@ -5,3 +5,9 @@ export const MAX_AVATAR_SIZE_MB_UNITS = 1024 * 1024
 export const COORD_ROUND_VALUE = 6
 export const OSM_BASE_URL = 'https://www.openstreetmap.org'
 export const OSM_DEFAULT_ZOOM = 18
+
+export const HttpStatus = {
```
Comment (Antonov2504) [inline packages/client/src/pages/profile/consts.ts]: Может быть стоит вынести в общие константы? Выглядит переиспользуемой в других компонентах и тестах

File: packages/client/src/pages/profile/PasswordSection.test.tsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+describe('PasswordSection', () => {
+  afterEach(() => {
+    jest.clearAllMocks()
+    jest.restoreAllMocks()
+  })
+
+  it('renders old password input', () => {
+    render(<PasswordSectionWrapper onChangePassword={jest.fn()} />)
+
+    expect(screen.getByLabelText(OLD_PASSWORD_LABEL)).toBeInTheDocument()
```
Comment (Antonov2504) [inline packages/client/src/pages/profile/PasswordSection.test.tsx]: На такие кейсы можно делать снапшоты. На мой взгляд, лаконичнее, быстрее и поддерживать проще

File: packages/client/src/pages/profile/ProfilePage.test.tsx [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+  beforeEach(() => {
+    jest.clearAllMocks()
+    jest.restoreAllMocks()
+    ;(profileService.getUser as jest.Mock).mockResolvedValue(FAKE_USER)
+  })
+
+  it('shows loading indicator while fetching user data', () => {
+    ;(profileService.getUser as jest.Mock).mockReturnValue(
+      new Promise(() => {
+        /* test */
```
Comment (Antonov2504) [inline packages/client/src/pages/profile/ProfilePage.test.tsx]: Здесь что-то предполагалось?

File: packages/client/src/pages/profile/ProfilePage.test.tsx [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+    await waitFor(() => {
+      expect(
+        screen.getByText('Не удалось загрузить данные профиля')
+      ).toBeInTheDocument()
+    })
+  })
+
+  it('shows error message when getUser rejects with a generic error', async () => {
+    ;(profileService.getUser as jest.Mock).mockRejectedValue(
```
Comment (Antonov2504) [inline packages/client/src/pages/profile/ProfilePage.test.tsx:120]: Не совсем понял, для чего ; в начале строки

File: packages/client/src/game/engine/Game.test.ts (hunk 1 above)
Comment (IrinaSha) [reply packages/client/src/game/engine/Game.test.ts]: ок, убрала

File: packages/client/src/pages/profile/ProfilePage.test.tsx [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    let resolveGetUser: (v: UserProfile) => void
+    ;(profileService.getUser as jest.Mock).mockImplementation(
+      () =>
+        new Promise<UserProfile>(resolve => {
+          resolveGetUser = resolve
+        })
+    )
+
+    const consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => {
+      /* test */
```
Comment (Antonov2504) [inline packages/client/src/pages/profile/ProfilePage.test.tsx]: И тут что-то должно быть?

File: packages/client/src/game/entities/Bullet.test.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { Bullet } from './Bullet'
+
+describe('Bullet', () => {
+  it('check update', () => {
+    const bullet = new Bullet({ x: 0, y: 100, vy: -200, owner: 'player' })
+    bullet.update(0.5)
+    // 100 + (-200)*0.5
```
Comment (Antonov2504) [inline packages/client/src/game/entities/Bullet.test.ts:7]: Комментированный код можно удалить, если не нужен

File: packages/client/src/pages/profile/ProfilePage.test.tsx (hunk 4 above)
Comment (IrinaSha) [reply packages/client/src/pages/profile/ProfilePage.test.tsx]: линтер ругалсо

File: packages/client/src/game/entities/Bullet.test.ts (hunk 7 above)
Comment (IrinaSha) [reply packages/client/src/game/entities/Bullet.test.ts:7]: согласна, но тут был, чтобы не забыть как считать)))

File: packages/client/src/pages/profile/consts.ts (hunk 2 above)
Comment (IrinaSha) [reply packages/client/src/pages/profile/consts.ts]: вынесла в shared

File: packages/client/src/pages/profile/ProfilePage.test.tsx (hunk 5 above)
Comment (IrinaSha) [reply packages/client/src/pages/profile/ProfilePage.test.tsx:120]: я так поняала (хотя мне привычнее другой стиль) что раз уж мы пишем без  ;  в конце строки то в случаях случае например с type assertion (profileService.getUser as jest.Mock) в начале нужно ставить

File: packages/client/src/pages/profile/ProfilePage.test.tsx (hunk 6 above)
Comment (IrinaSha) [reply packages/client/src/pages/profile/ProfilePage.test.tsx]: убрала)

File: packages/client/src/pages/profile/ProfilePage.test.tsx (hunk 4 above)
Comment (IrinaSha) [reply packages/client/src/pages/profile/ProfilePage.test.tsx]: переписала чуть чуть)
тут нужен был промис, но тело пустое и линтер ругался

File: packages/client/src/pages/profile/PasswordSection.test.tsx (hunk 3 above)
Comment (IrinaSha) [reply packages/client/src/pages/profile/PasswordSection.test.tsx]: переделала на снепшоты

**Review by Antonov2504 — APPROVED** [review]

File: packages/client/src/pages/profile/ProfileService.test.ts [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      )
+
+      const callArgs = (global.fetch as jest.Mock).mock.calls[0][1]
+
+      expect(callArgs.body).toBeInstanceOf(FormData)
+      expect(callArgs.body.get('avatar')).toEqual(file)
+      expect(result).toEqual(mockUser)
+    })
+  })
+})
```
Comment (MikeDevice) [inline packages/client/src/pages/profile/ProfileService.test.ts:192]: Это уже очень крепкий уровень тестов. Видно, что ты умеешь тестировать HTTP-слой, ошибки и edge-case'ы.

File: packages/client/src/pages/profile/ProfileService.test.ts [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+        ok: false,
+        status: HttpStatus.ServerError,
+        json: () => Promise.resolve({ reason: 'Error 500' }),
+      }) as jest.Mock
+
+      await expect(profileService.getUser()).rejects.toThrow('Error 500')
+    })
+
+    it('test unknown error invalid json', async () => {
+      global.fetch = jest.fn().mockResolvedValue({
```
Comment (MikeDevice) [inline packages/client/src/pages/profile/ProfileService.test.ts:96]: Ты мутируешь глобалку. Если запускать параллельно тесты, то будут конфликты

File: packages/client/src/pages/profile/GeoSection.test.tsx [hunk 10]
```diff
@@ -0,0 +1,86 @@
+import { render, screen } from '@testing-library/react'
```
Comment (MikeDevice) [inline packages/client/src/pages/profile/GeoSection.test.tsx:1]: Главная проблема — злоупотребление snapshot’ами. У тебя почти все тесты — snapshot. 

Проблема вот в чём:
изменила div → упало 6 тестов
обновил snapshot → ничего не проверяешь

Лучше вместо
```
expect(container).toMatchSnapshot()
```

вот так
```
expect(screen.getByRole('button', { name: BUTTON_NAME })).toBeInTheDocument()
```

File: packages/client/src/game/entities/Bullet.test.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    expect(bullet.isOffscreen(600)).toBe(true)
+  })
+
+  it('check offscreen false', () => {
+    const bullet = new Bullet({ x: 0, y: 300, vy: -1, owner: 'player' })
+    // y + h = 300 + 14 = 314 > 0 && 314 < screenH
+    expect(bullet.isOffscreen(600)).toBe(false)
+  })
+
+  it('check offscreen true', () => {
```
Comment (MikeDevice) [inline packages/client/src/game/entities/Bullet.test.ts:23]: Дублирующее название теста

File: packages/client/src/game/entities/Bullet.test.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    const bullet = new Bullet({ x: 0, y: 100, vy: -200, owner: 'player' })
+    bullet.update(0.5)
+    // 100 + (-200)*0.5
+    expect(bullet.y).toBe(0)
+  })
+
+  it('check offscreen true', () => {
+    const bullet = new Bullet({ x: 0, y: -20, vy: -1, owner: 'player' })
+    // y + h = -20 + 14 = -6 < 0
+    expect(bullet.isOffscreen(600)).toBe(true)
```
Comment (MikeDevice) [inline packages/client/src/game/entities/Bullet.test.ts:14]: тест знает внутренности класса, и если в Bullet поменяется (width, height), тесты упадут, хотя логика может быть ок

**Review by MikeDevice — APPROVED** [review]
Очень крутое покрытие тестами! Рекомендация: не злоупотреблять снапшотами и меньше зашивать в тесты знаний о системе. Тесты должны быть такой штукой, которая просто накладывается сверху и что-то проверяет, но сама не знает, как эта система устроена изнутри. Грубо говоря, тесты должны проверять, что `ЕСЛИ А, ТО Б`, без подробностей о том, как это `Б` было получено.
